numpy/core/src/multiarray/descriptor.c

```diff
@@ -509,8 +509,14 @@ _convert_from_array_descr(PyObject *obj, int align)
                  && (PyUString_Check(title) || PyUnicode_Check(title))
 #endif
                  && (PyDict_GetItem(fields, title) != NULL))) {
-            PyErr_SetString(PyExc_ValueError,
-                    "two fields with the same name");
+#if defined(NPY_PY3K)
+            name = PyUnicode_AsUTF8String(name);
+#else
+            Py_INCREF(name);
+#endif
+            PyErr_Format(PyExc_ValueError,
+                    "field '%s' occurs more than once", PyString_AsString(name));
+            Py_DECREF(name);
```

Comment:
Come to think of it, if we only put this `Py_DECREF` in an `#if defined(NPY_PY3K)` block, the line count of the code is going to be the same, and we spare ourselves a spurious `Py_INCREF` / `Py_DECREF` pair in Python 2... Sorry for the noise.